File: go/hashmap/290_word_pattern.go
package hashmap

import "strings"

// wordPattern 单词规律
func wordPattern(pattern string, s string) bool {
	// HashMap
	// 建立两个hashmap pattern 中的字符 -> s 中的string 的映射 和 反过来的映射

	hmap1 := make(map[byte]string)
	hmap2 := make(map[string]byte)

	vals := strings.Split(s, " ")

	if len(vals) != len(pattern) {
		return false
	}

	for i, v := range pattern {

		if val, ok := hmap1[byte(v)]; ok {
			if val != vals[i] {
				return false
			}
		}

		if val, ok := hmap2[vals[i]]; ok {
			if val != byte(v) {
				return false
			}
		}

		hmap1[byte(v)] = vals[i]
		hmap2[vals[i]] = byte(v)

	}
	return true
}

File: rust/src/hashmap/_290_word_pattern.rs
/// 单词规律
pub fn word_pattern(pattern: String, s: String) -> bool {
    // hashMap pattern -> s and s -> pattern

    use std::collections::HashMap;

    let (mut map1, mut map2) = (HashMap::new(), HashMap::new());

    if pattern.chars().collect::<Vec<_>>().len() != s.split(" ").collect::<Vec<_>>().len() {
        return false;
    }

    for (ch, word) in pattern.chars().zip(s.split(" ")) {
        if let Some(wd) = map1.insert(ch, word) {
            if wd != word {
                return false;
            }
        }

        if let Some(c) = map2.insert(word, ch) {
            if c != ch {
                return false;
            }
        }
    }
    true
}
